### ISSUE TITLE
API stability

### ISSUE THREAD

**Issue by davidmul87** [macxq/foxess-ha#20]
Hi all, 

Is anyone having issues with API stability? I was getting lots of issues over the weekend and once the integration has issues getting data it doesn't seem to automatically reconnect when the API comes back online? 

Regards,
David

**Comment by macxq** [comment]
Hi @davidmul87 

Didn't the integration connected even after next day, do you have any more datas - logs, screenshot describing the situation?

**Comment by davidmul87** [comment]
Hi Macxq, 

I will make sure to better document the next time it happens but it seems to me that when the integration wasn't able to connect to the API when it first boots then it doesn't then recover or retry connection and succeed when the API then comes back online until I restart my system again. As I say - apologies I don't have data to show this but if it next happens (considering the API stability I suspect that won't be long) then I will provide much better documentation. The integration did seem to tolerate the API outage last night so I wonder if this is just an issue with being unable to connect to the system when the integration first boots and then not recovering from that when it does come back online? 

I will either way send you some data records that I had from over the weekend as on one day the instability was also leading to some strange readings into the integration at the time which made the energy dashboard think my battery was getting rapidly charged and discharged as some of the figures were being read as zero and then returning again. See screenshot of this below (from Jan 9th). 

[url=https://ibb.co/8BZsN7p][img]https://i.ibb.co/pPFwbL8/energy-issues.png[/img][/url]
[url=https://ibb.co/LhhbpcW][img]https://i.ibb.co/M55WNt4/trace-of-the-data-series.png[/img][/url]

PS - thank you very much for your work putting the integration together - I had been searching for ages to try and find a way to monitor my solar system and battery - to the extent I was trying to find a hall-effect DC CT clamp I could hook onto the battery leads! This is much better if the FoxESS API would stay online... 

Regards,
David

**Comment by BigManDave** [comment]
@davidmul87 
I can confirm I've been seeing a lot of API connectivity issues over the past few days. However my install seems to drop the entities whilst the connection is unavailable and then reconnect without issue when it can.

**Comment by BigManDave** [comment]
Using entity history I can see the API drops. I've not been restarting when it reconnects:
![image](https://user-images.githubusercontent.com/66840216/149312019-b028d1d4-7c91-42be-a0a7-ac81df2ed9bc.png)

**Comment by Daibutt** [comment]
Ditto, my install also reconnects itself when API data is available again.
![image](https://user-images.githubusercontent.com/69276073/149314542-86e4f37b-7dfe-4c11-80fd-545d4434d1bf.png)

**Comment by macxq** [comment]
> The integration wasn't able to connect to the API when it first boots then it doesn't then recover or retry connection and succeed when the API then comes back online until I restart my system again.

That's can be true @davidmul87  and probably this could be fixed ;)

**Comment by davidmul87** [comment]
There was a bit of instability this morning in the API and i seem to have ended up with a data record that is making the data from the battery look strange - has anyone else seen this type of data?

![image](https://user-images.githubusercontent.com/63775737/149664542-b7a504b8-949a-4653-8ba9-12f17089884d.png)

![image](https://user-images.githubusercontent.com/63775737/149664575-d614aeaf-e64b-4004-ab30-45316029d960.png)

**Comment by BigManDave** [comment]
Mine looks fine today but I spotted similar on the 9th
![Screenshot_20220116-145338.jpg](https://user-images.githubusercontent.com/66840216/149665167-8ad162fb-8bff-4b15-9f3a-a400d8559de1.jpg)

**Comment by Daibutt** [comment]
I haven't seen this today, but have previously in the HA energy dashboard with both FoxESS data and my gas/elec imports.
I think the way the HA energy dashboard works out hourly usage is to see how much the energy totalisers (bat charge, bat discharge, load etc..) have increased in the previous hour. 
So if the totaliser drops to zero at some point in the day, when it returns to it's correct value, HA sees this as all being consumed in that particular hour.
If you look at your charts above @davidmul87, battery charge from 11:00-12:00 is suspiciously close to the sum of battery charge 00:00-04:00.

I had same issue with my gas/elec meters when I had connectivity issues with my meter app. When those dropped to zero & returned I had gas usage > 42,000kWh in one hour!

**Comment by davidmul87** [comment]
Hi all, 

Yep - that was what I was reading into it Daibutt. I don't understand fully the code (2 decades since I did any programming myself - but trying to get into it again) but I was wondering if it might need some safety tolerance before a zero is written to the data record? Either that or maybe checking with some sort of integral entity that is calculated off the daily power values to see if such a zero would make sense or not? The zeros really do mess with the data record. 

Regards,
David

**Comment by Daibutt** [comment]
Yes, for my gas/elec sensors I put in some logic that ignored a new data point that was less than roughly meter point's current value.
BUT.... that was an MQTT sensor and the logic was within the YAML for that. You could do the same with a template sensor for the FoxESS ones, but I'm not sure if template sensors can be used in the HA energy dashboard.
Also - gas/elec meters are always increasing, so any value lower than the current one is not correct. With the FoxESS ones they reset at midnight so you would have to allow them to go to zero then, but not at other times of the day.

**Comment by davidmul87** [comment]
Indeed – I was thinking that you could put some condition into the integration itself on whether to accept the value from the server or not to avoid erroneous zeros being recorded.

If (time >00:30am & time <23:30)
                If (value[t] > value [t-1])
                                Record new value
                End if
Else
                Record new value
End if

Obviously – I haven’t learnt either python or  other languages… You could potentially even be more specific to just ignore zeros that don’t make sense.

**Comment by macxq** [comment]
The energy meters in my case are rested when inverter in turned on again in the morning, but i do not have batteries:
<img width="1229" alt="Screenshot 2022-01-16 at 20 46 24" src="https://user-images.githubusercontent.com/2965092/149675429-330a8612-06f2-49d2-9a05-e48729b5be13.png">

I think the simplest solution is to ignore all the values if they are smaller than previous one for all the kWh entities in particular day.

**Comment by davidmul87** [comment]
Hi @macxq - I do think that would work well for me too - it may have some issues at some point depending on how well the API takes care of daylight savings. Is that something we should do via a template or would you update the integration code?

**Comment by macxq** [comment]
This is something that best would be done by integration code 👨‍💻  but i cannot promise when i will have time to take care of it

**Comment by pawel-matyaszczyk** [comment]
Hi @macxq!

PL:
Mam ten sam problem.  Rzeczywista produkcja 4,9 kW... a z wykresów, obliczeń prawie 22,4 kW. Generuje to kolejne problemy przy zliczaniu energii, obliczaniu kosztów, statystyki zużycia, wydajności, etc. 

Jest jakaś szansa na odrzucanie wyników totalnie błędnych? Zauważyłem w logach, że logujemy zawsze generation 0. 

ENG:
I have the same problem. Actual production 4.9 kW ... and from the graphs, calculations almost 22.4 kW. This generates further problems when counting energy, calculating costs, consumption and efficiency statistics, etc.

Any chance of rejecting totally wrong results? I noticed in the logs that we always log generation 0.

`2022-01-20 12:01:39 DEBUG (MainThread) [custom_components.foxess.sensor] FoxESS Earnings data fetched correcly {"errno":0,"result":{"currency":"PLN(Zł)","power":0.269,"today":{"generation":0,"earnings":0},"month":{"generation":70.5,"earnings":35.25},"year":{"generation":0,"earnings":0},"cumulate":{"generation":71.1,"earnings":35.55},"systemCapacity":4}}`

![Foxess](https://user-images.githubusercontent.com/58133061/150427869-5d093a27-6c72-4453-9193-742402cb33d5.png)

![Foxess](https://user-images.githubusercontent.com/58133061/150428094-40a60501-e63d-49a6-8252-bd2b851be8d8.png)


`

**Comment by BigManDave** [comment]
Greetings All,
I've got a possible fix for ignoring zeros on the energy sensors. I'm not a developer so might not be the best way to do it but it seems to work.
I've raised a PR for review, feel free to replicate the changes manually if you want to give it a test #34

**Comment by macxq** [comment]
v0.13 has been released with @BigManDave fix 👏

**Comment by williamjeccles** [comment]
I've not seen any 0 data entries messing up my data graphs since v0.13 - good job @BigManDave / @macxq
